## a. Why should the data be partitioned into training and validation sets? What will the training set be used for? What will the validation set be used for?

The data should be partitioned so that after the training, the model will have something that it hasn't seen before to predict on. As such, the training set is used to for learning and to fit the parameters, while the validation set is used to tune the hyperparameters.

## b. Fit a multiple linear regression model to the median house price (MEDV) as a function of CRIM, CHAS, and RM. Write the equation for predicting the median house price from the predictors in the model.

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import numpy as np

In [2]:
housing = pd.read_csv('3/BostonHousing.csv')

In [3]:
X = housing[['CRIM','CHAS','RM']]
y = housing['MEDV']

In [4]:
train_X, valid_X, train_y, valid_y = train_test_split(X,y, test_size=.3, random_state=10)

In [5]:
regr = LinearRegression()
regr.fit(train_X, train_y)

LinearRegression()

In [6]:
regr.intercept_

-27.87615474886597

In [7]:
regr.coef_

array([-0.29594781,  2.22379595,  8.11607882])

**Equation: MEDV = -27.876 + (-0.296 CRIM) + 2.224 CHAS + 8.116 RM**

## c. Using the estimated regression model, what median house price is predicted for a tract in the Boston area that does not bound the Charles River, has a crime rate of 0.1, and where the average number of rooms per house is 6?

In [8]:
regr.predict(np.array([[0.1,0,6]]))

array([20.79072341])

**MEDV = -27.876 -0.296 * 0.1 + 2.224 * 0 + 8.116 * 6 = 20.79 thousand**

## d. Reduce the number of predictors:
### i. Which predictors are likely to be measuring the same thing among the 13 predictors? Discuss the relationships among INDUS, NOX, and TAX.

Thinking logically, there's likely a positive relationship between INDUS, NOX and TAX because Boston being a well developed city, the higher the proportion of non-retail business (e.g. industrial businesses), the higher the polution as well as taxes in the region.

### ii. Compute the correlation table for the 12 numerical predictors and search for highly correlated pairs. These have potential redundancy and can cause multicollinearity. Choose which ones to remove based on this table.


In [9]:
housing.dtypes

CRIM         float64
ZN           float64
INDUS        float64
CHAS           int64
NOX          float64
RM           float64
AGE          float64
DIS          float64
RAD            int64
TAX            int64
PTRATIO      float64
LSTAT        float64
MEDV         float64
CAT. MEDV      int64
dtype: object

In [10]:
corr = housing.iloc[: , :12].corr()
corr.style.background_gradient(cmap='coolwarm')

From above, we see that the following variable pairs are highly correlated:
 NOX and INDUS: Correlation coefficient = 0.76365
 
 TAX and INDUS: Correlation coefficient = 0.72076
 
 AGE and NOX: Correlation coefficient = 0.73147
 
 DIS and NOX: Correlation coefficient = -0.76923
 
 DIS and INDUS: Correlation coefficient = -0.70803
 
 DIS and AGE: Correlation coefficient = -0.74788 
 
 TAX and RAD: Correlation coefficient = 0.91023
 
**Variables to be removed: TAX, INDUS, NOX, AGE**

Correlation Matrix after removing the four variables

In [11]:
corr = housing[['CRIM', 'ZN', 'CHAS', 'RM', 'DIS', 'RAD', 'PTRATIO', 'LSTAT']].corr()
corr.style.background_gradient(cmap='coolwarm')